Rebase rvh squash 0912

jueshiwenli opened the pull request.

shinezyy committed:
arch-riscv: add RISCV-H extension CSRs, WIP

jueshiwenli committed:
arch-riscv: fix bug of pmp

Change-Id: I5369768610a052299a3351b31677969d6a5d2f90

jueshiwenli committed:
arch-riscv: add mhpm & menvcfg & fix bugs

- add mcountinhibit
- add mhpmcounter
- update mcounteren config
- add menvcfg

jueshiwenli committed:
arch-riscv: make mideleg adapt h-ext

Change-Id: Ibffd33bff95044db9ec4f7e28c2c60ff874a0bb8

jueshiwenli committed:
arch-riscv: add hfence.gvma

Change-Id: Ib4b056c41e166ac374077b26567b31e9b681cfc6

jueshiwenli committed:
arch-riscv: make satp only adapt sv39

Change-Id: I88ccd869c7b33d64aef57d3e84fdc8e947d1c770

jueshiwenli committed:
dev: add plic read

Change-Id: I88f5910c6da05630eebd05cfcbe51ccffb376c92

jueshiwenli committed:
arch-riscv: update mstatus mask & mie mask

jueshiwenli committed:
arch-riscv: fix pmp bugs

make pmp read with different mask

Change-Id: If82e1f518e32c6e7da78d39664c124773930e9e8

jueshiwenli committed:
arch-riscv: add & fix bugs of h-csr & add h-diff

- add hstatus
- fix bug of vsstatus
- add h difftest
- add hie
- open virtualization mode
- fix bugs of mstatus & vir mode
- fix bugs of mastatus mask
- fix bugs of hstatus & virt mode
- fix bugs of write h csr

jueshiwenli committed:
arch-riscv: fix bugs of sret

- when cpu.v =1 sret need update vsstatus
- else need update v & hstatus

Change-Id: Ie7af67adae3c2fe337306517d0bec99ac4587056

jueshiwenli committed:
arch-riscv: fix bugs of pf

- when in s mode judge whether is in hs
- when in hs,update vsstatus v vscause vsepc

Change-Id: I49f790f565fba928c751590084420b24f3a1caff

jueshiwenli committed:
arch-riscv: fix bugs of hsmode

- fix bugs of write vsscratch
- fix bugs of pc state

Change-Id: I692366a494d2aeaff0e7f56035a6440e51ea550a

jueshiwenli committed:
arch-riscv: fix bugs of csr

- use g_trap update htval/mtval2 only g-stage
- when vmode need to update vsstatus instead of sstatus
- when write sstatus ,sometimes need to update vsstatus

Change-Id: I33edbb6ab81a82f65267878447e64e27b4e3a24a

jueshiwenli committed:
arch-riscv: support vmode ecall

Change-Id: I1baf3d38986371b3319cb3524d066c81929fe07b

jueshiwenli committed:
arch-riscv: add ptw tryCoalesce

Change-Id: I919f19814fe3ecf58de7ca551264dcdadc1d6bdc

jueshiwenli committed:
arch-riscv: fix bugs of marchid & sstatus

jueshiwenli committed:
arch-riscv: fix bugs of two stage ptw

when fault in vsstage need to endwalk and return
fix wrong of paddr

Change-Id: I707f14fb8ab57416dd6fb49d6c799ba70580fa24

jueshiwenli committed:
arch-riscv: fix H-csr bugs

- fix bugs of write vsstatus
- fix bugs of satp & hie & sie
- fix bugs of hcountern
- fix bugs of hip & hvip

Change-Id: I16e0bb4bec13c66b7ab814d2522db1009451dfdc

jueshiwenli committed:
arch-riscv: Vsatp mode 0 & misalign check

- when vsatp mode = 0 gstage need change translate level
- add misalign data addr check

Change-Id: Iabafd8726e8623f8ae0347f58490a43d7a879b29

jueshiwenli committed:
arch-riscv: support h-csr update when fault

- support h-csr update
- fix bugs of status vmode=1
- fix bugs of sepc stval scause

Change-Id: Id56a083364550f1165c4f4f2413cf7b6c491ee3c

jueshiwenli committed:
arch-riscv: fix bugs of h-ptw & sret

- when v = 1 , npc = vsepc
- initstate need set virt
- update pte.a & pte.d by pagefault

jueshiwenli committed:
arch-riscv: fix csr bugs & add wfi hvfault

- fix bugs of update vsstatus
- fix bugs of update mstatus
- rvh wfi need check hvfault
- fix bugs of update mstatus
- fix bugs of stvec satp stvec
- fix bugs of scountern
- fix bugs of read sie

Change-Id: I8e4d339eebf1bf80c2fe5337a9d9f4abbcc54262

jueshiwenli committed:
arch-riscv: fix bugs of h-ptw

- hgatp.mode == 8 || vsatp.mode == 8 need translate
- judge address before start g_ptw
- vsatpmode0, gpaddr == vaddr
- when coalesce, need to return different gpaddr

Change-Id: Ic8210e97a689e47bd457fc4cc2638550694f33b4

jueshiwenli committed:
arch-riscv: add tlb translate mode

Change-Id: I3b0f7858d4a27a23533628a3c38c50f41b5e9305

jueshiwenli committed:
arch-riscv: fix bugs of h-l1tlb

- all stage entry need put highest level in vs & g stage
- fix bugs of vsatp0mode gpaddr
- fix bugs of hit in l1tln gstage gPaddr
- allstsge hit need to check vsPte
- hfence.gvma will flush tlb with addr & hgatp
- hfence.vvma will flush tlb with addr & vsatp

Change-Id: I341eb4c17497c41660352e51cae57c0294e26f6b

jueshiwenli committed:
cpu: fix bugs of difftest

- add h-difftest when pagefault
- add stats of pagefault times

Change-Id: I51bcd1ebf3d0e7eda91ab335e775e829ec030382

jueshiwenli committed:
arch-riscv: clean h-l1tlb code

Change-Id: Ifdd38b47f6222c1ab6e285f46644b837dc80eb39

jueshiwenli committed:
arch-riscv: add h-l2tlb

- support hfence
- when hit in h-l2tlb the result will be send after set cycle
- check tlb before ptw send request

jueshiwenli committed:
arch-riscv: set misa with H-ext

Change-Id: I26add7336e515658b0bbdaebeb00bb2848826715

jueshiwenli committed:
dev: Add nemu plic

Change-Id: I6fe28f5654b5312eb7a3d77013f6f99fd006fa93

jueshiwenli committed:
sim: change gcpt-restore config

In h-checkpoint ,we don't need to designate gcpt-restore

Change-Id: I3433139c7e6b7f2774f3d9a1f7f0a5dbb874ea9e

jueshiwenli committed:
arch-riscv: add h inst

Change-Id: Idf45d80dde0450227a087ee14b1f2fa0680e3787

jueshiwenli committed:
arch-riscv: fix bugs of pmp

Change-Id: Ibf39612734e9ec64bfef3231c4732989a910ff9c

jueshiwenli committed:
arch-riscv: update csr mask

Change-Id: I842f48f4c78bd1f1b4dbe7b591a466beb924e167

jueshiwenli committed:
arch-riscv: add two stage translate mode

when need translate tlb will check translate mode
two stage translate mode :
- VSstage
- Gstage

Change-Id: I18a70651bf1a89a821609b417940fea5e5559bf7

jueshiwenli committed:
cpu: fix bugs of h-difftest

Change-Id: I6347d4ffd8dfca2c54b92e0573cf63dd0480367f

jueshiwenli committed:
arch-riscv: add h l1tlb

Change-Id: If3b3d3af7df6d460c7e6e106f87f5348927ab3d3

jueshiwenli committed:
arch-riscv: fix bugs of MTINST & MTVAL2

Change-Id: I27cdfd1370d1e702c69cc68ba395f4ef1de82dc8

jueshiwenli committed:
cpu: update difftest display

- default off h csr diff
- add switch to enable enabledifftesInstTrace

Change-Id: I91433e72310c8311dace7064ae0c0c3f6c71ed34

shinezyy commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -851,7 +851,8 @@ const RegVal MSTATUS_MASK = STATUS_SD_MASK | STATUS_SXL_MASK |
                             STATUS_SPP_MASK | STATUS_MPIE_MASK |
                             STATUS_SPIE_MASK | STATUS_UPIE_MASK |
                             STATUS_MIE_MASK | STATUS_SIE_MASK |
-                            STATUS_UIE_MASK;
```
Delete old code?

shinezyy commented on src/arch/riscv/isa/formats/standard.isa:
```diff
@@ -345,7 +345,8 @@ def template CSRExecute {{
         auto pm = (PrivilegeMode)xc->readMiscReg(MISCREG_PRV);
-        if (pm < lowestAllowedMode) {
+        if ((pm < lowestAllowedMode)&&
+            ((csr <CSR_HSTATUS) ||(csr >CSR_HGATP)) && ((csr <CSR_VSSTATUS) ||(csr > CSR_VSATP))) {
```
format thiese two line of code?

`csr <CSR_HSTATUS --> csr < CSR_HSTATUS`
`) ||( --> ) || (`

shinezyy commented on src/cpu/base.cc:
```diff
@@ -1114,6 +1114,232 @@ BaseCPU::diffWithNEMU(ThreadID tid, InstSeqNum seq)
+            diffAllStates->gem5RegFile.vsscratch = gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
```
Can we rewrite this chunk of code with a function, a template function, or a macro?

It looks too redundant

shinezyy commented on src/cpu/base.hh:
```diff
@@ -743,7 +743,7 @@ class BaseCPU : public ClockedObject
         // Register address causing difftest error
         bool errorRegsValue[96];// 32 regs + 32fprs +32 vprs
-        bool errorCsrsValue[32];// CsrRegIndex
+        bool errorCsrsValue[36];// CsrRegIndex
```
Can we use a enum or macro here?

shinezyy commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -4316,22 +4316,54 @@ decode QUADRANT default Unknown::unknown() {
-                                status.spp = PRV_U;
-                                xc->setMiscReg(MISCREG_STATUS, status);
-                                NPC = xc->readMiscReg(MISCREG_SEPC);
+                            bool writeSign = false;
```
local_variable should use snake_case

shinezyy commented on src/arch/riscv/faults.cc:
```diff
@@ -98,6 +100,18 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
+            if (pp != PRV_M && (tc->readMiscReg(MISCREG_VIRMODE) == 1) &&
+                bits(tc->readMiscReg(MISCREG_HEDELEG), _code) != 0) {
+                prvh = PRV_HS;
+            }
```
the if block and else block seems to be the same?

shinezyy replied:
redundancy has not resolved yet @jueshiwenli ?

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -979,7 +1259,15 @@ Walker::WalkerState::recvPacket(PacketPtr pkt)
+        if ((translateMode == twoStageMode) && (inGstage)){
+            mainFault = twoStageStepWalk(write);
+        }
+        else if ((translateMode == twoStageMode) && (!inGstage)){
```
} else if {

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -979,7 +1259,15 @@ Walker::WalkerState::recvPacket(PacketPtr pkt)
+        else if ((translateMode == twoStageMode) && (!inGstage)){
+            mainFault = twoStageWalk(write);
+        }
+        else{
```
} else if {

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1054,37 +1054,47 @@ TLB::checkPermissions(STATUS status, PrivilegeMode pmode, Addr vaddr,
+            code = ExceptionCode::STORE_PAGE;
+        else
+            code = ExceptionCode::INST_PAGE;
+    }
```
Are if block and else block the same?

shinezyy replied:
not resolved yet @jueshiwenli

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1255,6 +1264,44 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
+    }
+    if (virt != 0) {
+        if (vsatp.mode == 0)
+            assert(0);
```
`assert (vsatp.mode != 0);`

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1255,6 +1264,44 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
+        if (vsatp.mode == 0)
+            assert(0);
+    } else {
+        assert(0);
```
`asssert (virt != 0);`

shinezyy commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -1819,10 +1819,16 @@ LSQUnit::read(LSQRequest *request, ssize_t load_idx)
+        //DPRINTF(LSQUnit, "request: size: %u, Addr: %#lx, code: %d\n", request->mainReq()->getSize(),
+        //    request->mainReq()->getVaddr(), RiscvISA::ExceptionCode::LOAD_ADDR_MISALIGNED);
+        //return std::make_shared<RiscvISA::AddressFault>(request->mainReq()->getVaddr(),
+        //    RiscvISA::ExceptionCode::LOAD_ADDR_MISALIGNED);
```
remove them?

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -588,7 +590,7 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
             doEndWalk = true;
             assert(0);
-        } else if (((mode == BaseMMU::Execute) || (mode == BaseMMU::hLdST)) && (!pte.x)) {
+        } else if (((mode == BaseMMU::Execute) || (isHInst)) && (!pte.x)) {
```
redundant () ?

shinezyy commented on src/arch/riscv/faults.cc:
```diff
@@ -165,10 +169,13 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
                       _code == LOAD_ACCESS || _code == STORE_ACCESS)
                       tc->setMiscReg(MISCREG_HTVAL, 0);
+                  else if ((_code != INSTG_PAGE) && (_code != LOADG_PAGE) && (_code != STORE_PAGE))
+                      tc->setMiscReg(MISCREG_HTVAL, 0);
```
This code block is large, consider to add { } to avoid confusion

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -639,7 +632,11 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
+            assert(0);
         }
+    } else {
+        assert(0);
```
consider to change above two assertions into assert (xxx conditions)

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1603,14 +1633,19 @@ TLB::translate(const RequestPtr &req, ThreadContext *tc,
+                two_stage_translation = hasTwoStageTranslation(tc, req, mode);
+                if (two_stage_translation) {
+                    if (vsatp.mode != 8 && hgatp.mode != 8)
+                        assert(0);
```
change to assert (xx condition)

shinezyy commented on src/arch/riscv/isa.cc:
```diff
@@ -410,6 +410,15 @@ ISA::readMiscReg(int misc_reg)
+    }
+    if ((v == 1) && (misc_reg == MISCREG_SCAUSE)){
+        return readMiscRegNoEffect(MISCREG_VSCAUSE);
+    }
```
Can we rewrite above code like:

if (v == 1) and miscreg in [MISCREG_SATP, MISCREG_SPEC ....]:
   return ....

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -615,7 +620,12 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
             vaddr_choose_flag = (PgBase & 0x3f) / 8;
             nextRead = (PgBase >> 6) << 6;
             gPaddr = nextRead;
+            if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
```
change assertion into assert(condition)

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -697,12 +709,21 @@ Walker::WalkerState::twoStageWalk(PacketPtr &write)
                         gPaddr = gPaddr | (entry.vaddr & PGMASK);
+                        if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
+                            // this is a excep
+                            assert(0);
```
change assertion into assert(condition)

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -712,18 +733,28 @@ Walker::WalkerState::twoStageWalk(PacketPtr &write)
                     gPaddr = (pte.ppn << PageShift) + (idx_f * l2tlbLineSize);
+                    if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
+                        // this is a excep
+                        assert(0);
```
change assertion into assert(condition)

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -1552,19 +1594,33 @@ Walker::WalkerState::retry()
+            if (r.req->getPC() < page_start) {
+                vaddr = page_start;
+            } else
+                vaddr = r.req->getPC();
```
Either  use { } or not use { }. Don't mix them up in a code block.

XS-GEM5 always expected { } even if a block only has one line

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -1129,6 +1153,28 @@ Walker::WalkerState::endWalk()
 }
+Fault
+Walker::WalkerState::startTwoStageWalkFromTLB(Addr ppn, Addr vaddr)
+{
```
assert hgatp.mode == 8 ?

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -637,10 +642,16 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
+            if (twoStageLevel < level)
+                put_level = twoStageLevel;
+            else
+                put_level = level;
```
std::max?

shinezyy commented on src/cpu/o3/commit.cc:
```diff
@@ -1484,6 +1490,13 @@ Commit::commitHead(const DynInstPtr &head_inst, unsigned inst_num)
+            inst_fault->exception() == RiscvISA::ExceptionCode::INST_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::INSTG_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::LOADG_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::STOREG_PAGE)
```
put them in a set, then check whether exception is in this set?

shinezyy commented on src/cpu/o3/commit.cc:
```diff
@@ -1559,7 +1572,10 @@ Commit::commitHead(const DynInstPtr &head_inst, unsigned inst_num)
+                exception_no == RiscvISA::ExceptionCode::INST_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::INSTG_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::LOADG_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::STOREG_PAGE) {
```
simliar to above

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1082,6 +1079,26 @@ TLB::checkPermissions(STATUS status, PrivilegeMode pmode, Addr vaddr,
+    bool continuePtw = false;
+    if (pte.v && !pte.r && !pte.w && !pte.x) {
+        assert(0);
+        continuePtw = true;
```
???

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
+            assert(0);
+        if (vsatp.mode != 0) {
+            if ((mode == BaseMMU::Write && !e[0]->pteVS.d) || (!e[0]->pteVS.a))
+                fault = createPagefault(vaddr, 0, mode, false);
```
add  { }

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
+                req->setPaddr(gPaddr);
+                return std::make_pair(hit_type,NoFault);
+            }
+            else{
```
} else {

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
+                return std::make_pair(hit_type,NoFault);
+            }
+        }
+        else{
```
} else {

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -626,11 +679,45 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
+                                                     gstage);
+                        }
+
+                        else if (l2_level == 1)  // hit level =1
```
add { }

shinezyy commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -626,11 +679,45 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
+                                                     gstage);
+                        else if (l2_level == 2)  //
+                            walker->tlb->L2TLBInsert(inl2Entry.gpaddr, inl2Entry, l2_level, L_L2sp1, l2_i, false,
+                                                     gstage);
```
add { }

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -646,7 +654,12 @@ TLB::L2TLBInsertIn(Addr vpn, const TlbEntry &entry, int choose, EntryList *List,
+                        lookupL2TLB(vpn + step * i, entry.vmid, BaseMMU::Read, true, choose, false, translateMode);
+                else
+                    newEntry =
+                        lookupL2TLB(vpn + step * i, entry.asid, BaseMMU::Read, true, choose, false, translateMode);
```
add { }

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
+            for (i = 0; i < l2TlbSpSize * l2tlbLineSize; i = i + l2tlbLineSize) {
+                if (tlbL2Sp[i].trieHandle)
+                    l2TLBRemove(i, L_L2sp1);
+            }
```
There should be a function, and call this function 4 times

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
+            for (i = 0; i < l2TlbSpSize * l2tlbLineSize; i = i + l2tlbLineSize) {
+                if (tlbL2Sp[i].trieHandle)
+                    l2TLBRemove(i, L_L2sp1);
+            }
```
There should be a function, and call this function 4 times

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
+                    DPRINTF(TLBVerbose3, "l2l3 remove vaddr %#x vpn %#x vpnl2l3\n", tlbL2L3[i].vaddr, vpn, vpnl2l3);
+                }
+            }
         }
```
There should be a function, and call this function x times.

shinezyy commented on src/arch/riscv/tlb.cc:
```diff
@@ -1415,6 +1725,13 @@ TLB::doTwoStageTranslate(const RequestPtr &req, ThreadContext *tc,
+    if (isStage2)
+        l2tlb = this;
+    else
+        l2tlb = static_cast<TLB *>(nextLevel());
```
add  { }

jueshiwenli committed:
cpu: update difftest

- fix bug of difftest, when valuediff need panic
- use DiffValue flag enable difftest trace
- clean difftest code

Change-Id: Ie5409ea04a892d7f7a878345db37df482d2e68f1

jueshiwenli committed:
arch-riscv: clean H-MMU code

Change-Id: Ifcb9a9982b93db26e3e37df4332f433452f8e762

jueshiwenli committed:
arch-riscv: fix bugs of l2tlb

- Only check misalignDataAddr in two stage translate
- fix bugs of remove l2tlb

jueshiwenli committed:
misc: update ref & restore path
